layer/unittest/trace_event_log_tests.cc

```diff
@@ -27,8 +31,167 @@ using ::testing::IsEmpty;
 
 namespace performancelayers {
 namespace {
-// An `InstantEvent`. Adds a `scope` variable to the args indicating event's
-// visibility in the trace viewer.
+std::string ValueToJsonString(bool value) { return value ? "true" : "false"; }
+
+std::string ValueToJsonString(int64_t value) { return std::to_string(value); }
+
+std::string ValueToJsonString(const std::vector<int64_t> &values) {
+  std::stringstream json_str;
+  json_str << "[";
+  size_t e = values.size();
+  for (size_t i = 0; i != e; ++i) {
+    const char *delimiter = i < e - 1 ? "," : "";
+    json_str << "\"0x" << std::hex << values[i] << "\"" << delimiter;
+  }
+  json_str << "]";
+  return json_str.str();
+}
+
+// Converts the duration to milliseconds, the defualt time unit in the `Trace
+// Event` format.
+std::string ValueToJsonString(Duration value) {
+  return std::to_string(value.ToMilliseconds());
+}
+
+// Converts the timestamp to milliseconds, the defualt time unit in the `Trace
+// Event` format.
+std::string ValueToJsonString(TimestampClock::time_point value) {
+  return std::to_string(ToUnixMillis(value));
+}
+
+// Appends all given attributes, including the type-specific attribute, to the
+// stream.
+void TraceArgsToJsonString(const std::vector<Attribute *> &args,
+                           std::ostringstream &json_str) {
+  json_str << ", " << std::quoted("args") << " : { ";
+  for (size_t i = 0, e = args.size(); i < e; ++i) {
+    json_str << std::quoted(args[i]->GetName()) << " : ";
+    switch (args[i]->GetValueType()) {
+      case kHashAttribute: {
+        json_str << "\"0x" << std::hex << args[i]->cast<HashAttr>()->GetValue()
+                 << "\"";
+        break;
+      }
+      case ValueType::kTimestamp: {
+        json_str << ValueToJsonString(
+            args[i]->cast<TimestampAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kDuration: {
+        json_str << ValueToJsonString(
+            args[i]->cast<DurationAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kBool: {
+        json_str << ValueToJsonString(args[i]->cast<BoolAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kInt64: {
+        json_str << ValueToJsonString(args[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::kString: {
+        json_str << std::quoted(args[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case ValueType::kVectorInt64: {
+        json_str << ValueToJsonString(
+            args[i]->cast<VectorInt64Attr>()->GetValue());
+        break;
+      }
+      case ValueType::kTraceEvent:
+        break;
+    }
+  }
+  json_str << " }";
+}
+
+// Appends the start timestamp and the duration of a `TraceEventAttr` to the
+// given stream.
+void CompleteEventToString(TimestampAttr timestamp,
```

Comment:
This function name is a little bit misleading because `Complete` is a verb. Since this does not return a `string`, can we call it `AppendCompleteEvent` or `PrintCompleteEvent`?

Reply:
Sure, I was a bit iffy about the name too.

Reply:
I'd think it obvious it's the stream it appends to, but up to you